### ISSUE TITLE
Rewrite scanner

### ISSUE THREAD

**Issue by unreal** [onerobotics/tp_plus#3]
The current scanner depends on rexical and uses lots of regular expressions. This is slow and unnecessary.
1. Write a benchmark to test current speed
2. Re-write the scanner by hand
3. Test benchmark and verify step 2 wasn't unnecessary

**Comment by unreal** [comment]
So I re-wrote the scanner, and it's now almost 4x faster!

```
Running benchmark for 60 seconds (with 5 seconds warmup.)

Calculating -------------------------------------
               scan:    14.000  i/100ms
        old scanner:     3.000  i/100ms
-------------------------------------------------
               scan:    149.849  (± 2.7%) i/s -      8.988k
        old scanner:     38.780  (± 2.6%) i/s -      2.328k

Comparison:
               scan::      149.8 i/s
        old scanner::       38.8 i/s - 3.86x slower
```

After doing all the work, I was a little worried when my initial benchmarks actually showed it being slower. However, I was able to fix the issues by profiling.
1. The new scanner uses 0 regular expressions. Ruby doesn't have an easy way of testing whether a character is alphabetical or numeric, so I used a simple switch statement on the character for both, which ended up being faster. (I tried `char.match(/^[[:alpha:]]$/)` and `char.match(/^[[:digit:]]$/)` as well, confirming my suspicious that regex would be slower)
2. My initial findings that the new scanner was slower was pretty much entirely due to the way that I did token lookup. I made two changes to speed things up:
   1. I initially defined the tokens as a class method when a simple constant would do
   2. I initially iterated over the entire hash of constants before returning nil when a simple `hash[key] || default_value` would do. Ruby simply returns `nil` when the key is not found, so it's unnecessary (and slow) to validate it yourself.

I've added two new rake tasks: benchmark and profile that exercise the scanner and parser. Hopefully things don't slow down as new features are added.

**Comment by unreal** [comment]
I should have squashed all the commits on the development branch, but you can see a few commits with the last one just merged 0649f78dcb3f14be6d46d376812bbf73727952f7

A nice side-effect is that the test suite runs faster now too.

**Comment by groupsixtech** [comment]
Awesome work! But things are a little broken on my end after I merged. Dont know if you are getting these errors or if its just me. DIV statements are broken. You might be wanting to get away from this logic, but I was actually using it instead of a modulus for some things. namespaces are broken, and pr components are broken. 

RuntimeError: Parse error on line 3: 
parse error on value 2 (DIGIT)
C:/Users/Matt/Documents/tp_plus-Master/tp_plus/lib/tp_plus/parser.rb:35:in `rescue in parse'
    C:/Users/Matt/Documents/tp_plus-Master/tp_plus/lib/tp_plus/parser.rb:32:in`parse'
    C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:14:in `parse'
    C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:1242:in`test_div_statement'
    (eval):12:in `run'
    C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:93:in`start_mediator'
    C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:81:in `start'

<"IF (R[1:foo]<>1 OR R[2:foo2]<>2),JMP LBL[100] ;\nR[1:foo]=1 ;\nR[2:foo2]=2 ;\nLBL[100] ;\n"> expected but was
<"IF (R[1:foo]<>1 OR R[2:foo2]<>2),JMP LBL[100] ;\n ;\nR[1:foo]=1 ;\nR[2:foo2]=2 ;\nLBL[100] ;\n">.

diff:
  IF (R[1:foo]<>1 OR R[2:foo2]<>2),JMP LBL[100] ;
-  ;
  R[1:foo]=1 ;
  R[2:foo2]=2 ;
  LBL[100] ;
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:26:in `assert_prog'
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:1238:in`test_if_statement_multiple_arguments'
  (eval):12:in `run'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:93:in`start_mediator'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:81:in `start'

<"PR[1,1:Fixture pick_offset]=10 ;\n"> expected but was
<"PR[1:Fixture pick_offset]=10 ;\n">.

diff:
- PR[1,1:Fixture pick_offset]=10 ;
  ?    --
- PR[1:Fixture pick_offset]=10 ;
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:26:in `assert_prog'
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:733:in`test_namespaced_position_reg_component'
  (eval):12:in `run'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:93:in`start_mediator'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:81:in `start'

<"PR[1:Foo bar]=P[1:baz] ;\nPR[1,1:Foo bar]=PR[1,1:Foo bar]+5 ;\n"> expected but was
<"PR[1:Foo bar]=P[1:baz] ;\nPR[1:Foo bar]=PR[1:Foo bar]+5 ;\n">.

diff:
  PR[1:Foo bar]=P[1:baz] ;
- PR[1,1:Foo bar]=PR[1,1:Foo bar]+5 ;
  ?    --               --
- PR[1:Foo bar]=PR[1:Foo bar]+5 ;
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:26:in `assert_prog'
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:1065:in`test_namespaced_pr_assignment_then_expression'
  (eval):12:in `run'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:93:in`start_mediator'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:81:in `start'

<"PR[1,1:Foo bar]=10 ;\n"> expected but was
<"PR[1:Foo bar]=10 ;\n">.

diff:
- PR[1,1:Foo bar]=10 ;
  ?    --
- PR[1:Foo bar]=10 ;
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:26:in `assert_prog'
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:1050:in`test_namespaced_pr_component_assignment'
  (eval):12:in `run'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:93:in`start_mediator'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:81:in `start'

<"PR[1,1:Foo bar]=(PR[1,1:Foo bar]+PR[1,1:Foo bar]_10) ;\n"> expected but was
<"PR[1:Foo bar]=(PR[1:Foo bar]+PR[1:Foo bar]_10) ;\n">.

diff:
- PR[1,1:Foo bar]=(PR[1,1:Foo bar]+PR[1,1:Foo bar]*10) ;
  ?    --                --              --
- PR[1:Foo bar]=(PR[1:Foo bar]+PR[1:Foo bar]*10) ;
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:26:in `assert_prog'
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:1060:in`test_namespaced_pr_component_in_expression'
  (eval):12:in `run'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:93:in`start_mediator'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:81:in `start'

<"PR[1,1:Foo bar]=PR[1,1:Foo bar]+10 ;\n"> expected but was
<"PR[1:Foo bar]=PR[1:Foo bar]+10 ;\n">.

diff:
- PR[1,1:Foo bar]=PR[1,1:Foo bar]+10 ;
  ?    --               --
- PR[1:Foo bar]=PR[1:Foo bar]+10 ;
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:26:in `assert_prog'
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:1055:in`test_namespaced_pr_component_plus_equals'
  (eval):12:in `run'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:93:in`start_mediator'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:81:in `start'

Notification: TestInterpreter#test_position_data_populates_interpreter_position_data was redefined
C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:1182:in `<class:TestInterpreter>'
    C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:3:in`<top (required)>'
    C:/Ruby21-x64/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:15:in `require'
    C:/Ruby21-x64/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:15:in`block in <main>'
    C:/Ruby21-x64/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:4:in `select'
    C:/Ruby21-x64/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:4:in`<main>'

<"PR[1,1:foo]=5 ;\nPR[1,2:foo]=6 ;\nPR[1,3:foo]=7 ;\nPR[1,4:foo]=8 ;\nPR[1,5:foo]=9 ;\nPR[1,6:foo]=10 ;\n"> expected but was
<"PR[1:foo]=5 ;\nPR[1:foo]=6 ;\nPR[1:foo]=7 ;\nPR[1:foo]=8 ;\nPR[1:foo]=9 ;\nPR[1:foo]=10 ;\n">.

diff:
- PR[1,1:foo]=5 ;
  ?    --
- PR[1:foo]=5 ;
- PR[1,2:foo]=6 ;
  ?     --
- PR[1:foo]=6 ;
- PR[1,3:foo]=7 ;
  ?     --
- PR[1:foo]=7 ;
- PR[1,4:foo]=8 ;
  ?     --
- PR[1:foo]=8 ;
- PR[1,5:foo]=9 ;
  ?     --
- PR[1:foo]=9 ;
- PR[1,6:foo]=10 ;
  ?     --
- PR[1:foo]=10 ;
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:26:in `assert_prog'
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:401:in`test_pr_components'
  (eval):12:in `run'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:93:in`start_mediator'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:81:in `start'

<"PR[GP1:1,2:foo]=5 ;\n"> expected but was
<"PR[1:foo]=5 ;\n">.

diff:
- PR[GP1:1,2:foo]=5 ;
  ?    -- ----
- PR[1:foo]=5 ;
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:26:in `assert_prog'
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:1232:in`test_pr_components_groups'
  (eval):12:in `run'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:93:in`start_mediator'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:81:in `start'

<RuntimeError> exception expected but was
<Test::Unit::AssertionFailedError(<<""> expected but was
<"PR[1:foo]=5 ;\n">.>)>.

diff:
- RuntimeError
- Test::Unit::AssertionFailedError(<<""> expected but was
- <"PR[1:foo]=5 ;\n">.>)
  C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:406:in `test_pr_with_invalid_component_raises_error'
  (eval):12:in`run'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:93:in `start_mediator'
  C:/Program Files (x86)/JetBrains/RubyMine 7.0.3/rb/testing/patch/testunit/test/unit/ui/teamcity/testrunner.rb:81:in`start'

Notification: TestInterpreter#test_simple_linear_motion was redefined
C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:1225:in `<class:TestInterpreter>'
    C:/Users/Matt/Documents/tp_plus-Master/tp_plus/test/tp_plus/test_interpreter.rb:3:in`<top (required)>'
    C:/Ruby21-x64/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:15:in `require'
    C:/Ruby21-x64/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:15:in`block in <main>'
    C:/Ruby21-x64/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:4:in `select'
    C:/Ruby21-x64/lib/ruby/gems/2.1.0/gems/rake-10.4.2/lib/rake/rake_test_loader.rb:4:in`<main>'

**Comment by unreal** [comment]
Yeah, I expected some breaking changes on your branch since I had to refactor the parser pretty heavily.  Sorry about that.

The DIV bug was always there, and I didn't catch it because I didn't have proper test coverage for it! Fixed in 86d8ffc351736ef96f40eac4bd7893d316a34b33

Probably the best way to move forward is to cherry-pick your changes for 1) term(-1) for FINE and 2) mixed logic conditionals, and maybe we can open up a fresh pull request against the latest master for multi-group support.

**Comment by groupsixtech** [comment]
Yeah. nm. Its all good now. Figured out the DIV for the scanner too. That scanner will be much easier to use now going forward I think. Thanks for doing that. I know you are just going to start from scratch again for the next iteration in golang  so I hope updating this repo isn't completely useless to you. I'll put up a new pull request tomorrow sometime so we can get that sorted.
